File: web/lib/tiers.ts
/**
 * Tier vocabulary shared by the UI (US-26.2).
 *
 * The authoritative table is the backend's `services/tiers.py` — this is what the
 * frontend needs to *explain* a lock, not to decide one. The API's 403 carries the same
 * `feature`/`message`/`upgrade_url` fields, so a lock discovered server-side and one
 * predicted client-side read the same way.
 */

/** Where an upgrade prompt sends someone. Mirrors the API's `upgrade_url`. */
export const UPGRADE_URL = "/settings/billing"

/** What is locked, and what upgrading would let them do. */
export type LockedFeature = {
  name: string
  benefit: string
}

/**
 * Copy for the features the UI can lock before the server is asked.
 *
 * Keyed by the same capability strings the API's 403 uses (`detail.capability`), so a
 * 403 that arrives anyway can be rendered through the same modal.
 */
export const LOCKED_FEATURES: Record<string, LockedFeature> = {
  stems: {
    name: "Stem separation",
    benefit: "split a track into vocals, drums, bass and other",
  },
  midi: {
    name: "MIDI extraction",
    benefit: "export melody, chords, drums and bass as MIDI",
  },
  mastering: {
    name: "Mastering",
    benefit: "master your tracks to a professional loudness target",
  },
  distribution: {
    name: "Distribution",
    benefit: "publish your releases to streaming platforms",
  },
  voice_models: {
    name: "Custom voice models",
    benefit: "train a voice and sing your songs in it",
  },
  studio_editing: {
    name: "Studio editing",
    benefit: "arrange and mix in the multi-track Studio",
  },
  high_res_video: {
    name: "1080p and 4K video",
    benefit: "render videos above 720p, without a watermark",
  },
  lossless_export: {
    name: "Lossless export",
    benefit: "download WAV and FLAC as well as MP3",
  },
}

/** What the upgrade modal lists. Deliberately short — a wall of copy sells nothing. */
export const PRO_BENEFITS = [
  "500 credits a month, instead of 50",
  "Stems, MIDI, and lossless WAV/FLAC export",
  "Mastering and distribution to streaming platforms",
  "Custom voice models trained on your own recordings",
  "Full multi-track Studio editing",
  "1080p and 4K video, without a watermark",
]

/**
 * The copy for a locked capability, or a generic fallback.
 *
 * Falls back rather than throwing: a capability the server knows about and this build
 * does not should still produce a usable prompt, not a crash in a dropdown.
 */
export function lockedFeature(
  capability: string | null | undefined
): LockedFeature | null {
  if (!capability) return null
  return (
    LOCKED_FEATURES[capability] ?? {
      name: "This feature",
      benefit: "unlock everything Auto Music Studio can do",
    }
  )
}

File: src/acemusic/api/services/tiers.py
"""Subscription tiers and what each one unlocks (US-26.2).

One place that answers "may this tier do this?", so the API and the UI cannot drift into
disagreeing — the frontend badges and the backend 403 both derive from this table.

**The UI gating is decoration; this is the enforcement.** Before this module no endpoint
checked the tier at all, so a free account that called ``POST /clips/{id}/stems`` directly
got stems. A Pro badge in a dropdown stops an honest user, not a curious one.

Kept free of I/O and of ``HTTPException`` like the other service modules: it is a pure
lookup, so the rules are testable without a request, a database, or a user.
"""

from enum import Enum

FREE = "free"
PRO = "pro"

#: Monthly credit allocation per tier (US-26.2). This is what a reset tops the balance
#: back up to on the subscription anniversary.
MONTHLY_CREDITS = {
    FREE: 50.0,
    PRO: 500.0,
}


class Capability(str, Enum):
    """A thing a musician might do that the tier table has an opinion about.

    Named for the *capability* rather than the endpoint, because several endpoints can
    gate on the same one (single and batch mastering) and one endpoint can gate on
    different ones depending on its arguments (video resolution, export format).
    """

    STEMS = "stems"
    MIDI = "midi"
    MASTERING = "mastering"
    DISTRIBUTION = "distribution"
    VOICE_MODELS = "voice_models"
    STUDIO_EDITING = "studio_editing"
    HIGH_RES_VIDEO = "high_res_video"
    LOSSLESS_EXPORT = "lossless_export"


#: What each capability is called when explaining the lock, and why it is worth having.
#: The 403 body carries these, so a musician who hits one is told what to do about it
#: rather than just refused.
_DESCRIPTIONS = {
    Capability.STEMS: ("Stem separation", "split a track into vocals, drums, bass and other"),
    Capability.MIDI: ("MIDI extraction", "export melody, chords, drums and bass as MIDI"),
    Capability.MASTERING: ("Mastering", "master your tracks to a professional loudness target"),
    Capability.DISTRIBUTION: ("Distribution", "publish your releases to streaming platforms"),
    Capability.VOICE_MODELS: ("Custom voice models", "train a voice and sing your songs in it"),
    Capability.STUDIO_EDITING: ("Studio editing", "arrange and mix in the multi-track Studio"),
    Capability.HIGH_RES_VIDEO: ("1080p and 4K video", "render videos above 720p, without a watermark"),
    Capability.LOSSLESS_EXPORT: ("Lossless export", "download WAV and FLAC as well as MP3"),
}

#: Everything in :class:`Capability` is Pro-only — the free tier is deliberately defined
#: by what it *lacks*, so adding a capability locks it by default rather than leaking it.
#: A future middle tier would turn this into a per-tier set.
_PRO_ONLY = frozenset(Capability)


def normalise(tier: str | None) -> str:
    """The tier to actually apply.

    Anything that is not exactly ``"pro"`` is free — the same rule the frontend uses. An
    unknown, missing or malformed tier must fail *closed*: a typo in a database field
    should not hand out Pro.
    """
    return PRO if tier == PRO else FREE


def allows(tier: str | None, capability: Capability) -> bool:
    """Whether ``tier`` may use ``capability``."""
    return normalise(tier) == PRO or capability not in _PRO_ONLY


def monthly_allocation(tier: str | None) -> float:
    """Credits this tier is topped up to each month."""
    return MONTHLY_CREDITS[normalise(tier)]


def describe(capability: Capability) -> tuple[str, str]:
    """``(name, benefit)`` for ``capability`` — what is locked, and what it would do."""
    return _DESCRIPTIONS[capability]
